### ISSUE TITLE
aws-vault refresh credentials

### ISSUE THREAD

**Issue by christophetd** [99designs/aws-vault#816]
I'm often in the situation where I have an aws-vault shell opened, but the associated (AWS SSO) credentials have expired. Exiting it and re-running `aws-vault exec` is painful.

Is there any way, currently, to refresh the credentials? e.g. `aws-vault refresh` that would make sure the current credentials are either still valid, either refreshed?

If not, happy to send out a pull request.

Potential behavior:
1. If $AWS_SESSION_EXPIRATION is in the future, nothing to do unless a `--force` flag is passed
2. Re-perform the equivalent of `aws-vault exec $AWS_VAULT` and export the required environment variables in the _same_ shell

Thanks!

**Comment by christophetd** [comment]
Note: _"export the required environment variables in the same shell"_ is actually not possible. Possibilities I see:

1. Spin up a new process - in this case the `refresh` command would be pointless as it would behave the same as `aws-vault exec`
2. Spit out the required environment variables, so that a user could do `eval $(aws-vault refresh)`. With a shell alias it would be useful and to too cumbersome to use on a daily basis (e.g. `alias refresh='eval $(aws-vault refresh)'`)

Thoughts?

**Comment by deadlysyn** [comment]
@christophetd would you mind detailing your use case a bit more? partially to help me understand, and also for community feedback on how common this is/if better options exist.

it sounds like you literally `aws-vault exec profile -- zsh` or similar and then run multiple commands that way. i've used aws-vault for years and just exec commands (so each gets fresh creds as needed) and create a small shell wrapper (`v $command`) to make that less painful.

aside from saving typing, what are the use cases for execing a shell?

your second implementation suggestion seems sane, it's simple and works similar to well-known things like ssh-agent. another option you've likely already considered that isn't perfect but might help in the short term is setting `--duration` to something longer than an hour so you have to refresh less often.

**Comment by Volatus** [comment]
@deadlysyn I also expected similar behavior where I'd be able to do something like `aws-vault refresh <profilename>` and it'd acquire new credentials and apply those instead. I'd be down to work on such a feature, but the project doesn't seem to include contributing guidelines and also I'm not sure what process to go through to get such a feature approved.

**Comment by christophetd** [comment]
Here's the use-case more precisely:

1. Use `aws-vault exec my-role` to retrieve temporary STS credentials
2. Do some work
3. At some point, the credentials expire. But I don't want to `exit && aws-vault exec my-role` because it makes me loose my current directory, environment variables, virtual environment, etc.

So the idea is that when temporary credentials expire, I can run `aws-vault refresh <profile-name>` and the following happens:

- If `<profile-name>` is omitted, use the current value of `AWS_VAULT` in the environment
- If the credentials associated to the profile are not temporary, exit with an error
- Otherwise, call the credentials provider to retrieve new credentials, and store them in the keychain
- If `<profile-name>` was omitted, export the refreshed credentials in the environment

Thoughts? @deadlysyn @mtibben Happy to send out a PR for this

**Comment by christophetd** [comment]
This has been merged and released as part of aws-vault  v1.2.1

**Comment by mtibben** [comment]
> export the refreshed credentials in the environment

To source into your environment, it would look something like `eval $(aws-vault refresh --export-env)`.

This could also become the new home for the `--json` flag which was always misplaced in `exec`

**Comment by christophetd** [comment]
not stale, waiting for input on https://github.com/99designs/aws-vault/pull/886

**Comment by mtibben** [comment]
Resolved by #1135. You can now refresh credentials in your shell with `eval $(aws-vault export --format=export-env PROFILE)`

**Comment by viatcheslavmogilevsky** [comment]
The running of `eval $(aws-vault export --format=export-env PROFILE)` gives `aws-vault: error: exec: in an existing aws-vault subshell; 'exit' from the subshell or unset AWS_VAULT to force`, so I'm using the following:

```
eval $(AWS_VAULT=  aws-vault export --format=export-env $AWS_VAULT)
```
